docs/topics/clusterdefinitions.md

```diff
@@ -105,7 +105,7 @@ $ aks-engine get-versions
 | container-monitoring                                                                                      | false                                                                                | 1                               | Delivers the Kubernetes container monitoring component                                                                                                                                                                         |
 | [blobfuse-flexvolume](https://github.com/Azure/kubernetes-volume-drivers/tree/master/flexvolume/blobfuse) | true                                                                                 | as many as linux agent nodes    | Access virtual filesystem backed by the Azure Blob storage                                                                                                                                                                     |
 | [smb-flexvolume](https://github.com/Azure/kubernetes-volume-drivers/tree/master/flexvolume/smb)           | false                                                                                | as many as linux agent nodes    | Access SMB server by using CIFS/SMB protocol                                                                                                                                                                                   |
-| [keyvault-flexvolume](../../examples/addons/keyvault-flexvolume/README.md)                                | true                                                                                 | as many as linux agent nodes    | Access secrets, keys, and certs in Azure Key Vault from pods                                                                                                                                                                   |
+| [keyvault-flexvolume](../../examples/addons/keyvault-flexvolume/README.md)                                | true                                                                                 | as many as linux agent nodes    | Access secrets, keys, and certs in Azure Key Vault from pods. This solution will be **DEPRECATED** in favor of csi-secrets-store addon                                                                                                                                                                  |
```

Comment:
enabled by default should be false now, right?

Reply:
we prob dont want to just remove it without some kind of deprecation message for awhile right?

Reply:
maybe output it as a warning message in the cli? wdyt?

Reply:
@jackfrancis how do we support removal of addon features without impacting users now?

Reply:
We normally only worry about this for upgrade scenarios, in which case we (unless there's a good reason to statically override) prefer the previous config even if a new cluster at the same version would produce a different default in a similar functional surface area.

And for new clusters, if we change a default between 1.15 and 1.16, we don't customarily warn users, as our position is if the user doesn't provide an explicit dissenting configuration, he/she assents to the platform defaults (obviously this means we have to be good about what we choose as defaults).

I think in this case we should make sure that we don't allow *both* addons to be present simultaneously, right?

And in terms of deprecation, yeah, if keyvault-flexvolume functionally works in versions beyond where we set csi-secrets-store as the default, we should probably still allow it with a log warning message at cluster creation time saying "this addon will be deprecated in a future release" or something like that. I think that matches the Kubernetes pattern as well, for example, in controller-manager pod logs.

Reply:
so sounds like we should:
- upgrade: keep flexvol and output deprecation message (for how many releases?)
- new clusters 1.16+: default to csi driver and output deprecation message for flexvol

WDYT?

Reply:
I like it (we only output the deprecation message if the user has explicitly chosen to enable keyvault-flexvol, I assume you mean)

Reply:
That sounds good to me!

Reply:
so for upgrading clusters with flexvol, how many releases should we wait until we remove support for it?

Reply:
Updated the PR!

- `keyvault-flexvolume` addon will only be enabled by default for <= 1.15
- `csi-secrets-store` addon will only be enabled by default for 1.16+ clusters
- When upgrading from <= 1.15 to 1.16+, 
    - If `keyvault-flexvolume` addon was enabled, it will still be enabled after upgrade.
    - If `csi-secrets-store` addon doesn't exist before upgrade to 1.16+, it will not be explicitly enabled during the upgrade.
- If `keyvault-flexvolume` addon is explicitly enabled for a new 1.16+ cluster, a warning message will be shown and the addon will be enabled.

```
WARN[0000] keyvault-flexvolume add-on will be DEPRECATED in favor of csi-secrets-store addon for 1.16+ 
```